### PR TITLE
Allowing optional attributes when responding to change password challenge

### PR DESCRIPTION
**Pull request by assyadh**
*Issue #, if available:* DOTNET-3350

*Description of changes:*
 Fixes https://github.com/aws/aws-sdk-net/issues/1129

There was a missing parameter when resetting password. 

Tested with my personal account, fixes the bug

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/Amazon.Extensions.CognitoAuthentication/CognitoUserAuthentication.cs [hunk 1]
```diff
@@ -173,18 +173,31 @@ public async Task<AuthFlowResponse> RespondToSmsMfaAuthAsync(RespondToSmsMfaRequ
         /// password required authentication challenge using an asynchronous call
         /// </summary>
         /// <param name="newPasswordRequest">RespondToNewPasswordRequiredRequest object containing the necessary 
+        /// <param name="requiredAttributes">Optional dictionnary of attributes that may be required by the user pool
+        /// Each attribute key must be prefixed by "userAttributes."
         /// parameters to respond to the current SMS MFA authentication challenge</param>
         /// <returns>Returns the AuthFlowResponse object that can be used to respond to the next challenge, 
         /// if one exists</returns>
-        public async Task<AuthFlowResponse> RespondToNewPasswordRequiredAsync(RespondToNewPasswordRequiredRequest newPasswordRequest)
```
Comment (sstevenkang) [inline src/Amazon.Extensions.CognitoAuthentication/CognitoUserAuthentication.cs]: isn't this source compatible, but binary incompatible change?  Seems to me we can keep `RespondToNewPasswordRequiredAsync(RespondToNewPasswordRequiredRequest newPasswordRequest)` and introduce the new function?

File: src/Amazon.Extensions.CognitoAuthentication/CognitoUserAuthentication.cs [hunk 2]
```diff
@@ -173,18 +173,31 @@ public async Task<AuthFlowResponse> RespondToSmsMfaAuthAsync(RespondToSmsMfaRequ
...
         {
+            var challengeResponses = new Dictionary<string, string>()
+            {
+                { CognitoConstants.ChlgParamNewPassword, newPasswordRequest.NewPassword},
+                { CognitoConstants.ChlgParamUsername, Username }
+            };
+
+            if (requiredAttributes != null)
+            {
+                foreach (var attributeName in requiredAttributes.Keys)
```
Comment (sstevenkang) [inline src/Amazon.Extensions.CognitoAuthentication/CognitoUserAuthentication.cs]:
```
foreach(KeyValuePair<string, string> attribute in requiredAttributes)
{
}
```

File: src/Amazon.Extensions.CognitoAuthentication/CognitoUserAuthentication.cs (hunk 1 above)
Comment (assyadh) [reply src/Amazon.Extensions.CognitoAuthentication/CognitoUserAuthentication.cs]: I have added a signature similar to the one already released. This should give 100% backward compatibility.

File: src/Amazon.Extensions.CognitoAuthentication/CognitoUserAuthentication.cs (hunk 2 above)
Comment (assyadh) [reply src/Amazon.Extensions.CognitoAuthentication/CognitoUserAuthentication.cs]: Done

File: src/Amazon.Extensions.CognitoAuthentication/CognitoUserAuthentication.cs [hunk 3]
```diff
@@ -178,13 +178,40 @@ public async Task<AuthFlowResponse> RespondToSmsMfaAuthAsync(RespondToSmsMfaRequ
         /// if one exists</returns>
         public async Task<AuthFlowResponse> RespondToNewPasswordRequiredAsync(RespondToNewPasswordRequiredRequest newPasswordRequest)
         {
+            var result = await RespondToNewPasswordRequiredAsync(newPasswordRequest, null).ConfigureAwait(false);
```
Comment (sstevenkang) [inline src/Amazon.Extensions.CognitoAuthentication/CognitoUserAuthentication.cs]: no need to await here, right?

Comment (assyadh) [reply src/Amazon.Extensions.CognitoAuthentication/CognitoUserAuthentication.cs]: The original method has the async keyword. Can that cause an issue if the async keyword is dropped?

Comment (matteo-prosperi) [reply src/Amazon.Extensions.CognitoAuthentication/CognitoUserAuthentication.cs]: You can remove both the `async` and `await` keywords from this method. `async` doesn't effect the signature of the method, it is just instructions for the compiler when building your method. You can safely remove the `async` keyword from the method as long as your method continues to return a `Task`.

Comment (assyadh) [reply src/Amazon.Extensions.CognitoAuthentication/CognitoUserAuthentication.cs]: Thanks for chiming in. fixed it.

**Pull request merged**